### PR TITLE
Petit MC functions

### PR DESCRIPTION
**Pull request by carmenromo**
In order to do the same analysis in MC and data for PETit, this PR includes the functions we use to process data adapted to the MC tables. The file of data functions has been changed too and a test has been fixed.

### CODE REVIEW

File: antea/reco/petit_data_reco_functions.py [hunk 1]
```diff
@@ -85,10 +85,10 @@ def select_contained_evts_in_det_plane(df: pd.DataFrame) -> pd.DataFrame:
     return df_cov_evts
 
 
-def compute_charge_percentage_in_corona(df: pd.DataFrame,
```
Comment (paolafer) [inline antea/reco/petit_data_reco_functions.py:88]: Why did you change `percentage` for `ratio`? You are returning a percentage (between 0 and 100) not a ratio (between 0 and 1), right?

**Comment by paolafer** [comment]
Since most of the functions for MC are the same as for data, except that MC dataframes don't have the `tofpet_id` column, I would propose to add that column to the MC and then use the same functions as data. I'm doing it in my scripts in the following way:

```
def tofpetid(sid):
    if sid < 100:
        return 0
    else:
        return 2
        
data['tofpet_id'] = data['sensor_id'].apply(tofpetid)
```

File: antea/reco/petit_data_reco_functions.py (hunk 1 above)
Comment (carmenromo) [reply antea/reco/petit_data_reco_functions.py:88]: You're right!

**Comment by carmenromo** [comment]
> Since most of the functions for MC are the same as for data, except that MC dataframes don't have the `tofpet_id` column, I would propose to add that column to the MC and then use the same functions as data. I'm doing it in my scripts in the following way:
> 
> ```
> def tofpetid(sid):
>     if sid < 100:
>         return 0
>     else:
>         return 2
>         
> data['tofpet_id'] = data['sensor_id'].apply(tofpetid)
> ```

I like the idea but there is also the difference in the `sensor_id` column. In data we have `evt_number` and `cluster`. Maybe instead of duplicating the functions (what doesn't fully convince me) I can add to the functions the option `mc` or `data` to be usable for both cases.

**Comment by paolafer** [comment]
>  I can add to the functions the option mc or data to be usable for both cases.

Ok!

File: antea/reco/petit_data_reco_functions.py (hunk 1 above)
Comment (carmenromo) [reply antea/reco/petit_data_reco_functions.py:88]: Yes, I have it in mind.

Comment (carmenromo) [reply antea/reco/petit_data_reco_functions.py]: Done, I have decided to use ratio, since it's a shorter word. I have change accordingly the code.

File: antea/reco/petit_reco_functions.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+def tofpetid(sid: int) -> int:
+    """
+    Returns 0 if sensor_id is below 100 (detector plane)
+    and 2 if not (coinc plane).
+    """
+    if sid < 100: return 0
+    else: return 2
+
+
+def params(df: pd.DataFrame, data_or_mc: str = 'mc'):
```
Comment (paolafer) [inline antea/reco/petit_reco_functions.py:13]: It seems simpler to me to use a `boolean` variable, such as `data`, which the code checks if it's true or false, since there are only two options. This way, you can remove the `ValueError`.

File: antea/reco/petit_reco_functions.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        raise ValueError(f"Unrecognized data_or_mc type: {data_or_mc}.")
+
+    return df, evt_groupby
+
+
+def compute_coincidences(df: pd.DataFrame, data_or_mc: str = 'mc') -> pd.DataFrame:
+    """
+    Returns the events in which both planes have detected charge.
+    """
+    df, evt_groupby = params(df, data_or_mc)
```
Comment (paolafer) [inline antea/reco/petit_reco_functions.py:30]: We don't want to add the `tofpet_id` column and identify the columns for the groupby in every function, do we? I would separate these two algorithms in two functions, use a name that describe what they do and than just use them in the script. In the functions that do the groupby we can simply pass them `evt_groupby` as an argument, instead of `data_or_mc`.

File: antea/reco/petit_reco_functions.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    df = df[df.tofpet_id == tofpet_id]
+    if len(df)==0:
+        return False
+
+    if tot_mode: # t2 - t1 instead of intg_w or efine_corrected
+        argmax = (df.t2 - df.t1).argmax()
+    else:
+        argmax = df[variable].argmax()
+
+    argmax = df[variable].argmax()
```
Comment (paolafer) [inline antea/reco/petit_reco_functions.py:65]: Why are you repeating this line? It makes line 61 without effect, right?

File: antea/reco/petit_reco_functions.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+                                          data_or_mc: str = 'mc',
+                                          det_plane: bool = True,
+                                          variable:   str = 'charge',
+                                          tot_mode:  bool = False) -> pd.DataFrame:
+    """
+    Returns a dataframe with only the events with maximum charge
+    at the central sensors. If MC is being analyzed, `variable`
+    should be `charge` and `tot_mode` False.
+    """
+    df, evt_groupby  = params(df, data_or_mc)
```
Comment (paolafer) [inline antea/reco/petit_reco_functions.py:79]: I would eliminate this line, see comment above.

File: antea/reco/petit_reco_functions.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+                                   variable: str = 'charge') -> pd.Series:
+    """
+    Computes the ratio of charge detected in the external corona of the detection
+    plane with respect to the total charge of the that plane.
+    """
+    df, evt_groupby = params(df, data_or_mc)
+
+    tot_ch_d = df[df.tofpet_id==0]          .groupby(evt_groupby)[variable].sum()
+    cor_ch   = df[df.sensor_id.isin(corona)].groupby(evt_groupby)[variable].sum()
+    return (cor_ch/tot_ch_d).fillna(0)
```
Comment (paolafer) [inline antea/reco/petit_reco_functions.py:129]: Why do you want to fill `NaN` with `0`? If `tot_ch_d` is zero, which would cause a `NaN` in the ratio, we'd better know it, I think, instead of treating the event like it is totally contained.

**Comment by paolafer** [comment]
The tests are failing because this import `import antea.reco.petit_data_reco_functions as drf` is wrong, please fix it with the correct name.

File: antea/reco/petit_reco_functions_test.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+@mark.parametrize("filename data_or_mc det_plane variable tot_mode".split(),
+                  (('petit_mc_test.pet.h5', 'mc',  True,          'charge', False),
+                   ('petit_mc_test.pet.h5', 'mc', False,          'charge', False),
+                   ('petit_data_test.h5', 'data',  True, 'efine_corrected', False),
+                   ('petit_data_test.h5', 'data', False, 'efine_corrected', False),
+                   ('petit_data_test.h5', 'data',  True,          'intg_w', False),
+                   ('petit_data_test.h5', 'data',  True,      'intg_w_ToT',  True),
+                   ('petit_data_test.h5', 'data', False,      'intg_w_ToT',  True)))
+def test_select_evts_with_max_charge_at_center(ANTEADATADIR, filename, data_or_mc,
```
Comment (paolafer) [inline antea/reco/petit_reco_functions_test.py:64]: This test has the same name as the previous one. What's the difference between the two tests? Isn't the former a special case of the latter?

File: antea/reco/petit_reco_functions.py (hunk 2 above)
Comment (carmenromo) [reply antea/reco/petit_reco_functions.py:13]: Yes, I thought about it,  I decided to use string because it seemed more visual to me, but it's true that we save the `ValueError`. I'll implement it.

File: antea/reco/petit_reco_functions.py (hunk 4 above)
Comment (carmenromo) [reply antea/reco/petit_reco_functions.py]: Fixed!

File: antea/reco/petit_reco_functions_test.py (hunk 7 above)
Comment (carmenromo) [reply antea/reco/petit_reco_functions_test.py]: It was a bug, I just fixed it!

File: antea/reco/petit_reco_functions.py (hunk 6 above)
Comment (carmenromo) [reply antea/reco/petit_reco_functions.py:126]: You're right, but NaN was breaking the code, so maybe we can fill it with another float, like `2`, since ratio goes from 0 to 1.

Comment (carmenromo) [reply antea/reco/petit_reco_functions.py:113]: I have been looking into this and now I remember why I was filling with 0 when it was `NaN`. The ratio results in `NaN` when `cor_ch ` is empty, so that means that there are no sensors with charge belonging to the corona. In this case the ratio should be zero as it is.

File: antea/reco/petit_reco_functions.py (hunk 3 above)
Comment (carmenromo) [reply antea/reco/petit_reco_functions.py]: I have deleted `params` and instead of `data/mc` in the functions, I have added `evt_groupby` as an argument. This way, the `df` must contain a column called `tofpet_id` and when using the function we should specify the `groupby` arguments. There are some examples in the tests and in the script `process_data_petit`.

File: antea/reco/petit_reco_functions.py (hunk 6 above)
Comment (paolafer) [reply antea/reco/petit_reco_functions.py:113]: But it can also be NaN if `tot_ch_d` is zero, right?

Comment (carmenromo) [reply antea/reco/petit_reco_functions.py:113]: Yes, but there are some checks in the functions ensuring that the charge in the detector plane is not zero. We can add it here too, because it's true that this function can be used without the others.

File: antea/scripts/process_data_petit.py [hunk 8]
```diff
@@ -2,7 +2,7 @@
 import numpy  as np
 import pandas as pd
 
-import antea.reco.data_reco_functions as drf
+import antea.reco.petit_reco_functions as prf
 
 
 def from_ToT_to_pes(x):
```
Comment (paolafer) [inline antea/scripts/process_data_petit.py:8]: I would put this function in the `petit_reco_functions.py` code, as we have the shaping function in the shaping_functions code.

Comment (carmenromo) [reply antea/scripts/process_data_petit.py]: Done!

**Review by paolafer — APPROVED** [review]
This PR adds functions used in data and MC analysis of PETit. It also adds scripts that illustrate how to use them. Approved!
